### PR TITLE
Map on HighLevelGraph Layers

### PR DESCRIPTION
**Pull request by madsbk**
This PR introduces two `HighLevelGraph` methods:

```python
    def map_basic_layers(
        self, func: Callable[[BasicLayer], Mapping]
    ) -> "HighLevelGraph":
        """Map `func` on each basic layer and returns a new high level graph.
        `func` should take a BasicLayer as input and return a new Mapping as output
        and **cannot** change the dependencies between Layers.
        If `func` returns a non-BasicLayer type, it will be wrapped in a `BasicLayer`
        object automatically.

        Parameters
        ----------
        func : callable
            The function to call on each BasicLayer

        Returns
        -------
        hlg : HighLevelGraph
            A high level graph containing the transformed BasicLayers and the other
            Layers untouched
        """

    def map_tasks(self, func: Callable[[Iterable], Iterable]) -> "HighLevelGraph":
        """Map `func` on all tasks and returns a new high level graph.
        `func` should take an iterable of the tasks as input and return a new
        iterable as output and **cannot** change the dependencies between Layers.

        Warning
        -------
        A layer is allowed to ignore the map on tasks that are part of its internals.
        For instance, Blockwise will only invoke `func` on the input literals.

        Parameters
        ----------
        func : callable
            The function to call on tasks

        Returns
        -------
        hlg : HighLevelGraph
            A high level graph containing the transformed tasks
        """
```

The motivation is to avoid materialization of high level graphs in Distributed: https://github.com/dask/distributed/issues/4119

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by mrocklin** [comment]
In principle this seems fine to me.  At first I was thinking that maybe we should make a `map_layers` method instead, or make these private at first.  I think that I'm inclined instead to mark layers generally as experimental and go ahead with things as they are.

File: dask/highlevelgraph.py
Comment (jrbourbeau) [inline dask/highlevelgraph.py:237]: Ah, I see you've already included this check! I should have looked here before opening up https://github.com/dask/dask/pull/6716 : )

File: dask/highlevelgraph.py
Comment (jrbourbeau) [inline dask/highlevelgraph.py:499]: Could you elaborate on why we always need to return a `BasicLayer`? We may want the flexibility to have `func` return `Layer`s of different types. Though we can certainly hold off on that until the need arises : )

File: dask/tests/test_highgraph.py
Comment (jrbourbeau) [inline dask/tests/test_highgraph.py:84]: Nitpick: I _think_ this should be equivalent for what we're trying to test

```suggestion
    assert isinstance(layers[0], BasicLayer)
    assert isinstance(layers[1], Blockwise)
```

File: dask/tests/test_highgraph.py
Comment (jrbourbeau) [inline dask/tests/test_highgraph.py:85]: I think this should work

```python
from dask.array.utils import assert_eq
assert_eq(y, [42] * 3)
```

and is more consistent with how we test dask arrays in other parts of the codebase

File: dask/tests/test_highgraph.py
Comment (jrbourbeau) [inline dask/tests/test_highgraph.py:109]: Same comment about using `assert_eq`

File: dask/tests/test_highgraph.py
Comment (jrbourbeau) [inline dask/tests/test_highgraph.py:106]: I'm a little confused as to what we're trying to test here

**Review by jrbourbeau — COMMENTED** [review]
Thanks for your work here @madsbk! Overall this looks good, I've left a few comments below

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py:90]: This is just because `Layer` isn't declared at this point. Python allows using strings like `"Layer"` in order to _delay_ the evaluation to after `class Layer` has been declared. Semantically, it is the same.

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py:237]: Yeah sorry for not mentioning  it earlier but I forgot all about that change :)

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py:503]: In `map_basic_layers()` we cannot assume that key dependencies doesn't changes, `func` is allowed to modify both keys and tasks.
As an optimization, we could allow the user to provide a new `key_dependencies` but let's save that for another PR.

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py:530]: Good point, `map_tasks()` cannot change the keys so in this case we can pass along `self.key_dependencies`.

File: dask/tests/test_highgraph.py
Comment (madsbk) [reply dask/tests/test_highgraph.py:85]: Fixed

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py:508]: No, you are absolutely right no need to restrict this to `BasicLayer` :)
Fixed

File: dask/tests/test_highgraph.py
Comment (madsbk) [reply dask/tests/test_highgraph.py:106]: Updated the description to more descriptive, hope it makes more sense?
```python 
    if use_layer_map_task:
        # In order to test the default map_tasks() implementation on a Blockwise Layer,
        # we overwrite Blockwise.map_tasks with Layer.map_tasks
        blockwise_layer = list(dsk.layers.values())[1]
        blockwise_layer.map_tasks = partial(Layer.map_tasks, blockwise_layer)
```

File: dask/highlevelgraph.py
Comment (jrbourbeau) [inline dask/highlevelgraph.py:511]: Since `HighLevelGraph.__init__` will cast layers which are not `Layer` types to a `BasicLayer`, this can be simplified to something like

```suggestion
        layers = {
            k: func(v) if isinstance(v, BasicLayer) else v
            for k, v in self.layers.items()
        }
```

File: dask/highlevelgraph.py
Comment (jrbourbeau) [inline dask/highlevelgraph.py:481]: API question: What do you think about having a more generic `map_layers` method where the `func` being mapped is responsible for checking the layer type? This would be to prevent us from needing to add `map_*_layers` methods as the number of layer types grows. 

Apologies for not bringing this point up earlier in the process

**Review by jrbourbeau — COMMENTED** [review]
Thanks for the updates @madsbk! Overall this looks good to me. I've left one small comment about a simplification we can make and another about a potential API change

File: dask/highlevelgraph.py
Comment (mrocklin) [reply dask/highlevelgraph.py:511]: @jrbourbeau due to long delays because of timezones, and the impressive backlog that @madsbk has built up, I would encourage you to push directly if you have high confidence in these things (assuming that @madsbk is ok with this)

File: dask/highlevelgraph.py
Comment (mrocklin) [reply dask/highlevelgraph.py:481]: My sense here is that this probably makes sense, but that it also probably doesn't need to be handled right now.  The API choices here don't preclude us adding intermediate methods like `map_layers` later.

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py:511]: That is fine with me!

Comment (madsbk) [reply dask/highlevelgraph.py:511]: ... to both comments :)

Comment (jrbourbeau) [reply dask/highlevelgraph.py:511]: Great, will do!

File: dask/highlevelgraph.py
Comment (jrbourbeau) [reply dask/highlevelgraph.py:481]: Sounds good, we can revisit this point later on as needed 👍

**Review by jrbourbeau — APPROVED** [review]
LGTM, thanks for all your work here @madsbk! Planning to merge this once CI passes

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py:481]: > API question: What do you think about having a more generic `map_layers` method where the `func` being mapped is responsible for checking the layer type? This would be to prevent us from needing to add `map_*_layers` methods as the number of layer types grows.

Make sense, my thinking behind `map_basic_layers()` was to have a map for the layers that are materialized low level graphs. And if the user wants to apply something on all layers, it is always possible to access the layers of a HLG directly. 

But I agree, let's wait a bit before we settle on the API.

**Pull request merged**
